lib/mixlib/shellout/windows/core_ext.rb

```diff
@@ -171,10 +180,19 @@ def create3(args)
       end
 
       env = nil
+      env_list = []
+
+      # Retrieve the environment variables for the specified user.
+      if hash["with_logon"]
+        logon, passwd, domain = format_creds_from_hash(hash)
+        logon_type = hash["elevated"] ? LOGON32_LOGON_BATCH : LOGON32_LOGON_INTERACTIVE
+        token = logon_user(logon, domain, passwd, logon_type)
+        env_list = retrieve_environment_variables(token, env_list)
+      end
 
       # The env string should be passed as a string of ';' separated paths.
       if hash["environment"]
-        env = hash["environment"]
+        env = env_list.any? ? merge_env_variables(env_list, hash["environment"]) : hash["environment"]
```

Comment:
```suggestion
        env = merge_env_variables(env_list, hash["environment"])
```

I don't think this is correct. Since `merge_env_variables` formats things into `"#{k}=#{v}"` style strings that probably always needs to happen. That function looks like it will do the right thing even if `env_list` is empty.

Reply:
If no `environment` is passed but we are spawning as a separate user (`with_logon`), then it seems the `env_list` created for the user above will not be merged into env and passed to `create_process_` function

Reply:
Hi @mwrock `environment` sets from here https://github.com/chef/mixlib-shellout/blob/7320274a6abd96ad6e0aebde835a236e7c66e6bb/lib/mixlib/shellout/windows.rb#L285-L299 and it will never be nil.

Reply:
Hi @phiggins if `env_list` empty then we don't need to pass it on `merge_env_variables(env_list, hash["environment"])`, rather then we can directly return `hash["environment"]` if I am correct.

Reply:
@kapilchouhan99 Thanks for the explanation!

Reply:
To clarify, this line of `merge_env_variables` converts key-value pairs into env style strings (ie, "foo=bar"): 

```
merged_env.map { |k, v| "#{k}=#{v}" }
```

In this conditional, you're calling `merge_env_variables` in one branch and not the other:

```
env = env_list.any? ? merge_env_variables(env_list, hash["environment"]) : hash["environment"]
```

That means that either:
* `hash["environment"]` needs to be converted into env strings, and that conversion is not happening in one branch of the conditional.
* `hash["environment"]` is already formatted as env strings, so the formatting is unnecessary calling `merge_env_variables` is not going to work.

Reply:
I left a longer comment explaining my reasoning, but I just reread the code and I was mistaken. This seems fine. 👍